Rework MMU tables init, allow more effective mappings, increase TA_RAM

lorc opened the pull request:
Hi,

Current implementation of MMU support has a few drawbacks:

 1) If you want to use sections (pgdirs, blocks) for mapping, you must ensure that region starts and ends on section boundary

 2) There is a code that can be moved into generic part (to `core_mmu.c`) from mmu-specific files (`core_mmu_v7,c` and `core_mmu_lpae.c`). Ideally mmu-specific files should contain only low-level primitives for page tables manipulation.

Due to 1. some secure RAM is wasted, because `TA_RAM` is aligned to `PGDIR_SIZE`, so there will be hole between OP-TEE and TA RAM. Also, as OP-TEE is mapped 1:1, it often does not start on PGDIR boundary, so large portions of it are mapped in small pages, while they can be mapped as pgdirs. 

Due to 2. you sometimes need to write code twice and spend twice much resources to maintain it.

This PR is a aimed to mitigate issues describes above. Basically, it moves page tables initialization to `core_mmu.c`, drops now unused code from `core_mmu_*.c`  and optimizes region mapping by using pgdirs (or L1, L2 blocks) when possible.

As result it reduces secure memory waste and simplifies pagetables initialization code. 

I tagged this PR as RFC because it was not tested with pager yet.

lorc commented:
Added pager support.

jenswi-linaro commented on core/arch/arm/mm/core_mmu.c:
```diff
@@ -1302,6 +1354,79 @@ static void set_pg_region(struct core_mmu_table_info *dir_info,
+			idx = core_mmu_va2idx(&tbl_info, vaddr);
+
+			/* VA / PA are aligned to block size at current level */
+			level_ok = !((vaddr | paddr)
```
Please move the level_ok logic into a separate function.

jenswi-linaro commented on core/arch/arm/mm/core_mmu.c:
```diff
@@ -1302,6 +1354,79 @@ static void set_pg_region(struct core_mmu_table_info *dir_info,
+		level = 1;
+
+		while (true) {
+			assert(level <= 3);
```
Please replace the magic number `3` with a define depending on `CFG_WITH_LPAE`

lorc commented:
Addressed сomments. I'll fix commit messages during squashing.

jenswi-linaro commented on core/arch/arm/mm/core_mmu_lpae.c:
```diff
@@ -610,20 +595,21 @@ bool core_mmu_find_table(vaddr_t va, unsigned max_level,
 	paddr_t pa;
 	uint64_t attr;
+	paddr_t block_size_on_next_lvl =
+		L1_XLAT_ADDRESS_SHIFT - (tbl_info->level) *
```
The alignment here looks a bit funny, why isn't `L1_XLAT_ADDRESS_SHIFT` on the line above?
Also there's the `()` aren't needed.

lorc commented on core/arch/arm/mm/core_mmu.c:
```diff
@@ -705,6 +715,43 @@ static void dump_mmap_table(struct tee_mmap_region *memory_map)
+				debug_print("%*s [LVL%d] VA:0x%010" PRIxVA
+					" TBL:0x%010" PRIxPA " %s\n",
+					level * 2, "", level, va, pa,
+					attr & TEE_MATTR_SECURE ? " S" : "NS");
```
No, actually, is shows not relevant information. I'll  put this under  `#ifndef CFG_WITH_LPAE`.

lorc commented:
Hi @etienne-lms,

Are you mentioned closed test from GP test suite? If yes, and if it will reproduce, I'll be grateful for publically available test case.

etienne-lms commented:
Well, the test has nothing specific, just an invalid MAC comparison request that should only crash the TA. I initially though the core was due to the stack unwind sequence, but it still occurs with `CFG_UNWIND=n`. Strangely disappears with debug switched on. Reproduced only with pager enable (could it be some missing `KEEP_PAGER()` use?).
I'll tell you when I found more info.

lorc commented:
Thanks!

In meantime, I addressed other your and Jens's comments.

etienne-lms commented on core/arch/arm/mm/core_mmu.c:
```diff
@@ -887,10 +936,12 @@ static void init_mem_map(struct tee_mmap_region *memory_map, size_t num_elems)
+			 * efficient pgdir mapping. Basically pa &
+			 * pgdir_mask should be == va & pgdir_mask
+			 */
+			va += (map->pa - va) & CORE_MMU_PGDIR_MASK;
```
Removing the old code (va roundup in case of non LPAE) is ok.
But regarding the added code, I feel this may over consume mmu entries. It seems useful only if the area spreads at least over a superblock. And even so, since it shifts the other areas, it is not obvious there is gain here. 
But maybe i failed to understand your change...

lorc replied:
Ah, yes, my first version checked for `map->size >= PGDIR_SIZE * 2`. But I didn't noticed any changes in tables usage. I can return it back, if you feel that it is needed.

etienne-lms replied:
You can simply force the case by hacking this way:
```patch
--- a/core/arch/arm/mm/core_mmu.c
+++ b/core/arch/arm/mm/core_mmu.c
@@ -898,7 +898,7 @@ static void init_mem_map(struct tee_mmap_region *memory_map, size_t num_elems)
 
        assert(!((va | end) & SMALL_PAGE_MASK));
 
-       if (core_mmu_place_tee_ram_at_top(va)) {
+       if (1 || core_mmu_place_tee_ram_at_top(va)) {
                /* Map non-flat mapped addresses below flat mapped addresses */
                for (map = memory_map; !core_mmap_is_end_of_table(map); map++) {
                        if (map->va)
```

etienne-lms commented on core/arch/arm/mm/core_mmu.c:
```diff
@@ -866,10 +913,12 @@ static void init_mem_map(struct tee_mmap_region *memory_map, size_t num_elems)
+			 * efficient pgdir mapping. Basically pa &
+			 * pgdir_mask should be == va & pgdir_mask
+			 */
+			va += ((map->pa - va) & CORE_MMU_PGDIR_MASK);
```
I think this will fail as this sequence assigns virtual addresses in decreasing order. Should rather be something like `va -= ((map->pa - va) & CORE_MMU_PGDIR_MASK);`, but please cross-check...

lorc replied:
Yeah, I thought about it. But `va = ROUNDDOWN(va, map->region_size);` at L913 moves `va` down. If I see it right, then `va` will be increased, but it will  not move over `map->region_size`.

I think, I'll try to simulate different cases to be sure.

etienne-lms replied:
You can simply force the case by hacking this way:
```patch
--- a/core/arch/arm/mm/core_mmu.c
+++ b/core/arch/arm/mm/core_mmu.c
@@ -898,7 +898,7 @@ static void init_mem_map(struct tee_mmap_region *memory_map, size_t num_elems)
 
        assert(!((va | end) & SMALL_PAGE_MASK));
 
-       if (core_mmu_place_tee_ram_at_top(va)) {
+       if (1 || core_mmu_place_tee_ram_at_top(va)) {
                /* Map non-flat mapped addresses below flat mapped addresses */
                for (map = memory_map; !core_mmap_is_end_of_table(map); map++) {
                        if (map->va)
```

lorc replied:
Yep, I tried similar test on qemu (v7 and v8). It worked perfectly  fine, but now I'm worried that I might miss some corner case...

etienne-lms commented:
Some news regarding the crash I found. It is not related to your changes but rather a missing `KEEP_PAGER()` constraint on `mobj_reg_shm_get_pa()`. It is used by the abort handler to dump the shm areas mapped in the TA space when a TA crashes or panics. Seems we did not faced it before because of bad luck.

lorc commented:
I did some synthetic testing and found that my approach was not quite right. In some cases it can lead to overlapping mappings. I hadn't any problems on qemu, but there might exist some cases, when we'll have a problem. So I reworked it in a right way. 

Also I added test `if (map->size > CORE_MMU_PGDIR_SIZE)` to reduce empty areas between mappings.

lorc commented:
Hi @etienne-lms,

The patch you mentioned is crucial for the next one: `"vexpress: align TA_RAM with small page size"`
Without right alignment, we can't effectively map whole TA RAM. Because, instead of using pgdir blocks, it will be mapped with small pages, which will waste pagetables.  If I'll drop `"mmu: align va of memory regions to pa modulo PGDIR_SIZE"`, then I'll be forced to drop `"vexpress: align TA_RAM with small page size"` too. 
But the latter one is really needed for virtualization. Actually, that patch is the reason, why this PR exists in the first place.

etienne-lms commented:
Your change has 2 parts: mmu to reuse tables for applying a given mapping and generic boot to change the optee core vmem layout to fulfil some needs. I think the 1st part you made is fine and valuable (factorization of existing code).

Regarding the 2nd part (vmem layout changes): 
Supporting virtualization requires quite a lot of memory. In such case, do we really need to handle case where memory is constrained? I mean, i op-tee executes in some external ram, their should be plenty of ddr for the required page table.
If optee executes in a small internal ram and has to handle virtualization of the REEs... then, i wonder how to dimension the TEE memory requirements. Is it a realistic usecase?

lorc commented:
Okay, I'll put  the first part in the separate PR.

As for virtualization use case.. I made this PR not because I need every bit of TA RAM. I just needed to ensure, that there are no holes in secure RAM mappings, because my implementation tries to use all available secure RAM. But right now there is 0-2M hole between TEE RAM and and TA RAM start thanks to pgdir alignment. I think, we can discuss this later, in more appropriate PR.

lorc commented:
Closing as a stale.
